File: frontend/src/game/upgrades/curses/damage_reduc_3.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const DamageReduc3Def: UpgradeDef = {
  id: "damage_reduc_3",
  name: "Weakness 3",
  description: "-0.8% damage.",
  rarity: RarityID.Rare,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 0,
  targetClass: UpgradeTargetID.Attack,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: -0.008,
  isMultiplier: true,
  stackable: true,
  maxStacks: 99999,
  curse: true,
}

export class DamageReduc3 extends Upgrade {}

File: backend/app/core/upgrades/curses/damage_reduc_3.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class DamageReduc3(UpgradeImplementation):
    id = 'damage_reduc_3'
    name = 'Weakness 3'
    description = '-0.8% damage.'
    rarity = 'rare'
    type = 'stat_modifier'
    cost = 0

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = -0.008
        appliers.modifiers.add_multiplier('attack', 'damage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True
